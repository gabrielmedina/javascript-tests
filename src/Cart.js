import { find, remove } from 'lodash'

export default class Cart {
  items = []

  add(item) {
    const itemToFind = { product: item.product }

    if(find(this.items, itemToFind)) {
      remove(this.items, itemToFind)
    }

    this.items.push(item)
  }

  remove(product) {
    remove(this.items, { product })
  }

  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc + item.quantity * item.product.price
    }, 0)
  }

  getSummary() {
    return {
      total: this.getTotal(),
      items: this.items
    }
  }

  getCheckout() {
    const { total, items } = this.getSummary()

    this.items = []

    return {
      total,
      items
    }
  }
}
