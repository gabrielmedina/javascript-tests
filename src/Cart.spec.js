import Cart from './Cart'

describe('Cart', () => {
  let cart
  let product

  beforeEach(() => {
    cart = new Cart()

    product = {
      title: 'A cool product',
      price: 5388,
    }
  })

  describe('getTotal()', () => {
    it('should multiply quantity and price and recive the total amount', () => {
      cart.add({
        product,
        quantity: 2,
      })

      expect(cart.getTotal()).toEqual(10776)
    })

    it('should ensure no more than on product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product,
        quantity: 1,
      })

      expect(cart.items.length === 1).toEqual(true)
      expect(cart.getTotal()).toEqual(5388)
    })

    it('should update total when a product gets included and then removed', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product: { ...product, price: 3276 },
        quantity: 1,
      })

      expect(cart.getTotal()).toEqual(14052)

      cart.remove(product)

      expect(cart.getTotal()).toEqual(3276)
    })
  })

  describe('getSummary()', () => {
    it('should return an object with the total and the list of items when getSummary() is called', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product: { ...product, price: 3276 },
        quantity: 2,
      })

      expect(cart.getSummary()).toMatchSnapshot()
      expect(cart.getTotal()).toBeGreaterThan(0)
    })
  })
  
  describe('getCheckout()', () => {
    it('should reset the cart when getCheckout() is called', () => {
      cart.add({
        product,
        quantity: 2,
      })
  
      cart.getCheckout()
  
      expect(cart.getTotal()).toEqual(0)
    })
  })
})
