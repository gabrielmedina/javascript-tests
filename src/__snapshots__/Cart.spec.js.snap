// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart getSummary() should return an object with the total and the list of items when getSummary() is called 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 5388,
        "title": "A cool product",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 3276,
        "title": "A cool product",
      },
      "quantity": 2,
    },
  ],
  "total": 17328,
}
`;
